nucypher/crypto/kits.py

```diff
@@ -55,11 +54,55 @@ def __init__(self,
                  ciphertext=None,
                  signature=NOT_SIGNED) -> None:
 
+        if type(capsule) == bytes:
+            raise Exception
+
         self.ciphertext = ciphertext
         self.capsule = capsule
         self.sender_verifying_key = sender_verifying_key
         self._signature = signature
 
+        self._cfrags = set()
+        self._delegating_key = None
+        self._receiving_key = None
+        self._verifying_key = None
+
+    def attach_cfrag(self, cfrag: VerifiedCapsuleFrag):
+        # TODO: check that the cfrag belongs to this capsule?
+        self._cfrags.add(cfrag)
+
+    def clear_cfrags(self):
+        self._cfrags = set()
+
+    def set_correctness_keys(self, delegating=None, receiving=None, verifying=None):
+
+        if self._delegating_key is None:
+            self._delegating_key = delegating
+        elif delegating is not None and delegating != self._delegating_key:
+            raise Exception("Replacing an existing delegating key")
+
+        if self._receiving_key is None:
+            self._receiving_key = receiving
+        elif receiving is not None and receiving != self._receiving_key:
+            raise Exception("Replacing an existing receiving key")
+
+        if self._verifying_key is None:
+            self._verifying_key = verifying
+        elif verifying is not None and verifying != self._verifying_key:
+            raise Exception("Replacing an existing verifying key")
+
+    def get_correctness_keys(self):
+        return dict(delegating=self._delegating_key,
+                    receiving=self._receiving_key,
+                    verifying=self._verifying_key)
+
+    def __len__(self):
+        # TODO: may be better to just have a "has enough cfrags" method
+        return len(self._cfrags)
+
+    def __str__(self):
+        return f"{self.__class__.__name__}({self.capsule}, {len(self)} cfrags"
```

Comment:
This is my favorite way to construct `__repr__` and `__str__`.

Missing closing parenthesis:

```suggestion
        return f"{self.__class__.__name__}({self.capsule}, {len(self)} cfrags)"
```

Reply:
Fixed